### ISSUE TITLE
Downloads: add download failure reasons

### ISSUE THREAD

**Issue by bychkovdmitry** [google/ExoPlayer#5914]
### [REQUIRED] Use case description
User has ran out of free space on his/her device. He/she starts download and after several retries gets status update with "unknown" failure reason. 

### Proposed solution
Add FAILURE_REASON_INSUFFICIENT_FREE_SPACE to Download.FailureReason IntDef.

### Alternatives considered
If aforementioned solution is hard to implement, could you at least add support for minimum storage requirement in bytes for the download to proceed. (This might be even harder as currently you store the combination of requirement flags in an Int and there is no way to store extra data)

**Comment by ojw28** [comment]
Generalizing this issue to also track adding other failure reasons (e.g. network failures).

**Comment by sravan1213** [comment]
@ojw28 Can this failure reason be customized?? So that app can decide what the failure reason is ??

Example - our server throws 403 for multiple reasons.. token expired.. geo blocked etc...

We can distinguish only if we are able to read http response headers and failure response..

**Comment by mseroczynski** [comment]
BTW, ExoPlayer can also throw errors when we try to free up some space - this happened when I tried to remove some files from full storage with ExoPlayer's api.

`2020-05-27 15:16:32.780 12546-12962/com.xxx.debug E/SQLiteLog: (13) statement aborts at 5: [DELETE FROM ExoPlayerCacheIndex18390749ed91b365 WHERE id = ?] database or disk is full
2020-05-27 15:16:32.787 12546-12962/com.xxx.debug E/SQLiteLog: (1) statement aborts at 1: [ROLLBACK;] cannot rollback - no transaction is active`

**Comment by sravan1213** [comment]
@ojw28 is there some way you can pass exception to consuming app without storing in DB. Let the app decide what it has to do with exception.

I would suggest changing 
`default void onDownloadChanged(DownloadManager downloadManager, Download download) {}`

to 

`default void onDownloadChanged(DownloadManager downloadManager, Download download, @Nullable Throwable error) {}`

If this change makes sense to you, I am happy to raise change request :)

**Comment by ojw28** [comment]
@sravan1213 - That sounds like a good idea to me. I assume that internally in `DownloadManager`, your proposal would be to add `finalError` into `DownloadUpdate`, and then forward it on to the listener in `onDownloadUpdate`? Please feel free to open a pull request into the `dev-v2` branch if so!

**Comment by sravan1213** [comment]
Yes, using DownloadUpdate is what I was thinking... Will make this change.. 👍

**Comment by sravan1213** [comment]
@ojw28 please have a look at https://github.com/google/ExoPlayer/pull/7479

**Comment by ojw28** [comment]
#7479 was merged. Thanks! This issue still needs to remain open for tracking the addition of failure reasons that can be persisted in the database.

**Comment by sravan1213** [comment]
@ojw28 could you please confirm when above change will be part of the exo-release ?

**Comment by ojw28** [comment]
It will be in 2.12, which should be released very soon now (we were originally aiming for last week, and now for this week!)
